src/ui/ui_element.h

```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include <vector>
+
+#include "render2d.h"
+
+class UIElement {
+public:
+    UIElement() {} // default ctor
+    virtual ~UIElement() {}
+    UIElement(float start_x, float start_y)
+        : start_x(start_x), start_y(start_y) {}
+    virtual void draw(Render2D &renderer_2d,
```

Comment:
Seems a bit cumbersome to pass in a `Render2D` instance to every call. There's already a lot of code with this call signature though, so it's not a priority to change this. I do think in the future that providing `Render2D` statically would be easier to work with. Example:

```c++
class UIElement {
public:
  static void provide_render2d(const Render2D *renderer_) {
    renderer = renderer_;
  }
  void draw(float offset_x, float offset_y) {
    assert(renderer);
    draw_impl(offset_x, offset_y);
  }
    
protected:
  const Render2D *renderer;

  virtual void draw_impl(float offset_x, float offset_y) = 0;
  // Can use renderer->render_rectP(...) in implementing classes.
}

// Use as...

UIElement::provide_render2d(render2d_inst_ptr);
UIElement e;
e.draw(x, y);
```

I have no idea what this pattern is even called, but the best name I've got for it is static dependency injection. (As always, open to criticism.)